# Project 1: Trading with Momentum
## Instructions
Each problem consists of a function to implement and instructions on how to implement the function.  The parts of the function that need to be implemented are marked with a `# TODO` comment. After implementing the function, run the cell to test it against the unit tests we've provided. For each problem, we provide one or more unit tests from our `project_tests` package. These unit tests won't tell you if your answer is correct, but will warn you of any major errors. Your code will be checked for the correct solution when you submit it to Udacity.

## Packages
When you implement the functions, you'll only need to you use the packages you've used in the classroom, like [Pandas](https://pandas.pydata.org/) and [Numpy](http://www.numpy.org/). These packages will be imported for you. We recommend you don't add any import statements, otherwise the grader might not be able to run your code.

The other packages that we're importing are `helper`, `project_helper`, and `project_tests`. These are custom packages built to help you solve the problems.  The `helper` and `project_helper` module contains utility functions and graph functions. The `project_tests` contains the unit tests for all the problems.

### Install Packages

In [1]:
import sys
!{sys.executable} -m pip install -r requirements.txt

    100% |████████████████████████████████| 880kB 9.4MB/s eta 0:00:01
    100% |████████████████████████████████| 17.0MB 1.7MB/s eta 0:00:01   15% |█████                           | 2.6MB 25.1MB/s eta 0:00:01    30% |█████████▋                      | 5.1MB 26.2MB/s eta 0:00:01    44% |██████████████▎                 | 7.6MB 24.2MB/s eta 0:00:01    51% |████████████████▋               | 8.8MB 25.5MB/s eta 0:00:01    66% |█████████████████████▏          | 11.3MB 25.2MB/s eta 0:00:01    87% |████████████████████████████    | 14.8MB 25.7MB/s eta 0:00:01
    100% |████████████████████████████████| 26.2MB 1.2MB/s eta 0:00:01  0% |▏                               | 133kB 21.8MB/s eta 0:00:02    4% |█▋                              | 1.3MB 25.6MB/s eta 0:00:01    13% |████▌                           | 3.6MB 24.6MB/s eta 0:00:01    52% |████████████████▉               | 13.8MB 18.1MB/s eta 0:00:01    71% |██████████████████████▉         | 18.7MB 24.1MB/s eta 0:00:01    79% |██████████████████████

    100% |████████████████████████████████| 81kB 5.7MB/s ta 0:00:01
  Running setup.py bdist_wheel for cvxpy ... done
  Stored in directory: /root/.cache/pip/wheels/2b/60/0b/0c2596528665e21d698d6f84a3406c52044c7b4ca6ac737cf3
  Running setup.py bdist_wheel for plotly ... done
  Stored in directory: /root/.cache/pip/wheels/98/54/81/dd92d5b0858fac680cd7bdb8800eb26c001dd9f5dc8b1bc0ba
  Running setup.py bdist_wheel for scs ... done
  Stored in directory: /root/.cache/pip/wheels/df/d0/79/37ea880586da03c620ca9ecd5e42adbd86bc6ea84363965c5f
Successfully built cvxpy plotly scs
tensorflow 1.3.0 requires tensorflow-tensorboard<0.2.0,>=0.1.0, which is not installed.
moviepy 0.2.3.2 has requirement tqdm==4.11.2, but you'll have tqdm 4.19.5 which is incompatible.
  Found existing installation: numpy 1.12.1
    Uninstalling numpy-1.12.1:
      Successfully uninstalled numpy-1.12.1
  Found existing installation: scipy 1.2.1
    Uninstalling scipy-1.2.1:
      Successfully uninstalled scipy-1.2.1
  Foun

### Load Packages

In [1]:
import pandas as pd
import helper
import project_helper
import project_tests
import numpy as np

In [2]:
df = pd.read_csv('../../data/project_1/eod-quotemedia.csv', parse_dates=['date'], index_col=False)
close = df.reset_index().pivot(index='date', columns='ticker', values='adj_close')
print('Loaded Data')

Loaded Data


### View Data
Run the cell below to see what the data looks like for `close`.

In [3]:
project_helper.print_dataframe(close)

AttributeError: module 'plotly.graph_objs' has no attribute 'Table'

## Market Data
### Load Data
The data we use for most of the projects is end of day data. This contains data for many stocks, but we'll be looking at stocks in the S&P 500. We also made things a little easier to run by narrowing down our range of time period instead of using all of the data.

## Resample Adjusted Prices

The trading signal you'll develop in this project does not need to be based on daily prices, for instance, you can use month-end prices to perform trading once a month. To do this, you must first resample the daily adjusted closing prices into monthly buckets, and select the last observation of each month.

Implement the `resample_prices` to resample `close_prices` at the sampling frequency of `freq`.

In [4]:
apple_ticker = 'AAPL'
project_helper.plot_stock(close[apple_ticker], '{} Stock'.format(apple_ticker))

### Stock Example
Let's see what a single stock looks like from the closing prices. For this example and future display examples in this project, we'll use Apple's stock (AAPL). If we tried to graph all the stocks, it would be too much information.

In [5]:
def resample_prices(close_prices, freq='M'):
    """
    Resample close prices for each ticker at specified frequency.
    
    Parameters
    ----------
    close_prices : DataFrame
        Close prices for each ticker and date
    freq : str
        What frequency to sample at
        For valid freq choices, see http://pandas.pydata.org/pandas-docs/stable/timeseries.html#offset-aliases
    
    Returns
    -------
    prices_resampled : DataFrame
        Resampled prices for each ticker and date
    """
    print(close_prices.resample(freq).last())
    print(close_prices)
    # TODO: Implement Function
    return close_prices.resample(freq).last()
    

project_tests.test_resample_prices(resample_prices)

                   DWCO         HSR          JPHC        YNCL         IMQ
2008-08-31  21.05081048 17.01384381   10.98450376 11.24809343 12.96171273
2008-09-30 482.34539247 35.20258059 3516.54167823 66.40531433 13.50396048
2008-10-31  10.91893302 17.90864387   24.80126542 12.48895419 10.52435923
2008-11-30  11.54549538 23.98146843   24.97476306 36.03196210 14.30433232
                   DWCO         HSR          JPHC         YNCL         IMQ
2008-08-19  21.05081048 17.01384381   10.98450376  11.24809343 12.96171273
2008-09-08  15.63570259 14.69054309   11.35302769 475.74195118 11.95964043
2008-09-28 482.34539247 35.20258059 3516.54167823  66.40531433 13.50396048
2008-10-18  10.91893302 17.90864387   24.80126542  12.48895419 10.52435923
2008-11-07  10.67597197 12.74940144   11.80525758  21.53903949 19.99766037
2008-11-27  11.54549538 23.98146843   24.97476306  36.03196210 14.30433232
Tests Passed


### View Data
Let's apply this function to `close` and view the results.

In [13]:
monthly_close = resample_prices(close)
project_helper.plot_resampled_prices(
    monthly_close.loc[:, apple_ticker],
    close.loc[:, apple_ticker],
    '{} Stock - Close Vs Monthly Close'.format(apple_ticker))

ticker               A         AAL          AAP         AAPL        ABBV  \
date                                                                       
2013-07-31 30.77861719 18.63139292  81.73270857  58.73000866 38.52144972   
2013-08-31 32.09288410 15.55986096  79.33492514  63.64994327 36.09056668   
2013-09-30 35.34697923 18.25587647  81.98212977  62.28266407 37.88620154   
2013-10-31 35.00902763 21.15409315  98.34285959  68.28583759 41.39637606   
2013-11-30 36.94707663 22.60801580 100.15741326  73.07037475 41.39637606   
2013-12-31 39.53485221 24.31228275 109.80605374  73.72082947 45.12162270   
2014-01-31 40.19849023 32.30404302 113.90344263  65.78133976 42.40046983   
2014-02-28 39.35511692 35.55851889 126.35434591  69.56208595 43.84740848   
2014-03-31 38.65691442 35.24077420 125.56160936  70.95004943 44.26943225   
2014-04-30 37.44602782 33.76759430 120.39025770  78.00222579 45.26058952   
2014-05-31 39.45552968 38.66856535 123.24889355  84.14255587 47.21597213   
2014-06-30 3

## Compute Log Returns

Compute log returns ($R_t$) from prices ($P_t$) as your primary momentum indicator:

$$R_t = log_e(P_t) - log_e(P_{t-1})$$

Implement the `compute_log_returns` function below, such that it accepts a dataframe (like one returned by `resample_prices`), and produces a similar dataframe of log returns. Use Numpy's [log function](https://docs.scipy.org/doc/numpy/reference/generated/numpy.log.html) to help you calculate the log returns.

In [14]:
def compute_log_returns(prices):
    """
    Compute log returns for each ticker.
    
    Parameters
    ----------
    prices : DataFrame
        Prices for each ticker and date
    
    Returns
    -------
    log_returns : DataFrame
        Log returns for each ticker and date
    """
    # TODO: Implement Function
   
    print (np.log(prices))
    print(np.log(prices.shift(1)))
    
    return np.log(prices)- np.log(prices.shift(1))

project_tests.test_compute_log_returns(compute_log_returns)

                  JBT       ZZDO        FGK       FGGL       PXTR
2008-08-31 3.04693906 2.83402735 2.39648553 2.42019864 2.56199984
2008-09-30 6.17866044 3.56111939 8.16523331 4.19577709 2.60298301
2008-10-31 2.39049826 2.88528349 3.21089468 2.52484459 2.35369250
2008-11-30 2.44629535 3.17728138 3.21786584 3.58440638 2.66056245
                  JBT       ZZDO        FGK       FGGL       PXTR
2008-08-31        nan        nan        nan        nan        nan
2008-09-30 3.04693906 2.83402735 2.39648553 2.42019864 2.56199984
2008-10-31 6.17866044 3.56111939 8.16523331 4.19577709 2.60298301
2008-11-30 2.39049826 2.88528349 3.21089468 2.52484459 2.35369250
Tests Passed


### View Data
Using the same data returned from `resample_prices`, we'll generate the log returns.

In [15]:
monthly_close_returns = compute_log_returns(monthly_close)
project_helper.plot_returns(
    monthly_close_returns.loc[:, apple_ticker],
    'Log Returns of {} Stock (Monthly)'.format(apple_ticker))

ticker              A        AAL        AAP       AAPL       ABBV        ABC  \
date                                                                           
2013-07-31 3.42682020 2.92484795 4.40345427 4.07295082 3.65121522 3.98671202   
2013-08-31 3.46863433 2.74469458 4.37367845 4.15339843 3.58603152 3.97061867   
2013-09-30 3.56521294 2.90448703 4.40650129 4.13168312 3.63458697 4.04148377   
2013-10-31 3.55560596 3.05183342 4.58845994 4.22370239 3.72319334 4.10842325   
2013-11-30 3.60948653 3.11830452 4.60674308 4.29142301 3.72319334 4.18839927   
2013-12-31 3.67718262 3.19098169 4.69871566 4.30028538 3.80936157 4.18527516   
2014-01-31 3.69382944 3.47519239 4.73535110 4.18633621 3.74715944 4.14033195   
2014-02-28 3.67262600 3.57117976 4.83909023 4.24221968 3.78071562 4.15311601   
2014-03-31 3.65472566 3.56220377 4.83279655 4.26197610 3.79029442 4.11923987   
2014-04-30 3.62290064 3.51950159 4.79073861 4.35673736 3.81243667 4.11296930   
2014-05-31 3.67517421 3.65502701 4.81420

## Shift Returns
Implement the `shift_returns` function to shift the log returns to the previous or future returns in the time series. For example, the parameter `shift_n` is 2 and `returns` is the following:

```
                           Returns
               A         B         C         D
2013-07-08     0.015     0.082     0.096     0.020     ...
2013-07-09     0.037     0.095     0.027     0.063     ...
2013-07-10     0.094     0.001     0.093     0.019     ...
2013-07-11     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...
```

the output of the `shift_returns` function would be:
```
                        Shift Returns
               A         B         C         D
2013-07-08     NaN       NaN       NaN       NaN       ...
2013-07-09     NaN       NaN       NaN       NaN       ...
2013-07-10     0.015     0.082     0.096     0.020     ...
2013-07-11     0.037     0.095     0.027     0.063     ...
...            ...       ...       ...       ...
```
Using the same `returns` data as above, the `shift_returns` function should generate the following with `shift_n` as -2:
```
                        Shift Returns
               A         B         C         D
2013-07-08     0.094     0.001     0.093     0.019     ...
2013-07-09     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...       ...
...            ...       ...       ...       ...       ...
...            NaN       NaN       NaN       NaN       ...
...            NaN       NaN       NaN       NaN       ...
```
_Note: The "..." represents data points we're not showing._

In [16]:
def shift_returns(returns, shift_n):
    """
    Generate shifted returns
    
    Parameters
    ----------
    returns : DataFrame
        Returns for each ticker and date
    shift_n : int
        Number of periods to move, can be positive or negative
    
    Returns
    -------
    shifted_returns : DataFrame
        Shifted returns for each ticker and date
    """
    # TODO: Implement Function
    
    return returns.shift(shift_n)

project_tests.test_shift_returns(shift_returns)

Tests Passed


### View Data
Let's get the previous month's and next month's returns.

In [17]:
prev_returns = shift_returns(monthly_close_returns, 1)
lookahead_returns = shift_returns(monthly_close_returns, -1)

project_helper.plot_shifted_returns(
    prev_returns.loc[:, apple_ticker],
    monthly_close_returns.loc[:, apple_ticker],
    'Previous Returns of {} Stock'.format(apple_ticker))
project_helper.plot_shifted_returns(
    lookahead_returns.loc[:, apple_ticker],
    monthly_close_returns.loc[:, apple_ticker],
    'Lookahead Returns of {} Stock'.format(apple_ticker))

## Generate Trading Signal

A trading signal is a sequence of trading actions, or results that can be used to take trading actions. A common form is to produce a "long" and "short" portfolio of stocks on each date (e.g. end of each month, or whatever frequency you desire to trade at). This signal can be interpreted as rebalancing your portfolio on each of those dates, entering long ("buy") and short ("sell") positions as indicated.

Here's a strategy that we will try:
> For each month-end observation period, rank the stocks by _previous_ returns, from the highest to the lowest. Select the top performing stocks for the long portfolio, and the bottom performing stocks for the short portfolio.

Implement the `get_top_n` function to get the top performing stock for each month. Get the top performing stocks from `prev_returns` by assigning them a value of 1. For all other stocks, give them a value of 0. For example, using the following `prev_returns`:

```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0.015     0.082     0.096     0.020     0.075     0.043     0.074
2013-07-09     0.037     0.095     0.027     0.063     0.024     0.086     0.025
...            ...       ...       ...       ...       ...       ...       ...
```

The function `get_top_n` with `top_n` set to 3 should return the following:
```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0         1         1         0         1         0         0
2013-07-09     0         1         0         1         0         1         0
...            ...       ...       ...       ...       ...       ...       ...
```
*Note: You may have to use Panda's [`DataFrame.iterrows`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.iterrows.html) with [`Series.nlargest`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.Series.nlargest.html) in order to implement the function. This is one of those cases where creating a vecorization solution is too difficult.*

In [18]:
def get_top_n(prev_returns, top_n):
    """
    Select the top performing stocks
    
    Parameters
    ----------
    prev_returns : DataFrame
        Previous shifted returns for each ticker and date
    top_n : int
        The number of top performing stocks to get
    
    Returns
    -------
    top_stocks : DataFrame
        Top stocks for each ticker and date marked with a 1
    """
    # TODO: Implement Function
    mod_returns = prev_returns.copy()
    for index, row in prev_returns.iterrows():
        
        mod_returns.loc[index][row.index] = 0
        
        mod_returns.loc[index][row.nlargest(top_n).index] = 1
    
    mod_returns = mod_returns[mod_returns.columns].astype(np.int64)

    return mod_returns

project_tests.test_get_top_n(get_top_n)

Tests Passed


### View Data
We want to get the best performing and worst performing stocks. To get the best performing stocks, we'll use the `get_top_n` function. To get the worst performing stocks, we'll also use the `get_top_n` function. However, we pass in `-1*prev_returns` instead of just `prev_returns`. Multiplying by negative one will flip all the positive returns to negative and negative returns to positive. Thus, it will return the worst performing stocks.

In [19]:
top_bottom_n = 50
df_long = get_top_n(prev_returns, top_bottom_n)
df_short = get_top_n(-1*prev_returns, top_bottom_n)
project_helper.print_top(df_long, 'Longed Stocks')
project_helper.print_top(df_short, 'Shorted Stocks')

10 Most Longed Stocks:
INCY, AMD, AVGO, NFX, SWKS, NFLX, ILMN, UAL, NVDA, MU
10 Most Shorted Stocks:
RRC, FCX, CHK, MRO, GPS, WYNN, DVN, FTI, SPLS, TRIP


## Projected Returns
It's now time to check if your trading signal has the potential to become profitable!

We'll start by computing the net returns this portfolio would return. For simplicity, we'll assume every stock gets an equal dollar amount of investment. This makes it easier to compute a portfolio's returns as the simple arithmetic average of the individual stock returns.

Implement the `portfolio_returns` function to compute the expected portfolio returns. Using `df_long` to indicate which stocks to long and `df_short` to indicate which stocks to short, calculate the returns using `lookahead_returns`. To help with calculation, we've provided you with `n_stocks` as the number of stocks we're investing in a single period.

In [20]:
def portfolio_returns(df_long, df_short, lookahead_returns, n_stocks):
    """
    Compute expected returns for the portfolio, assuming equal investment in each long/short stock.
    
    Parameters
    ----------
    df_long : DataFrame
        Top stocks for each ticker and date marked with a 1
    df_short : DataFrame
        Bottom stocks for each ticker and date marked with a 1
    lookahead_returns : DataFrame
        Lookahead returns for each ticker and date
    n_stocks: int
        The number number of stocks chosen for each month
    
    Returns
    -------
    portfolio_returns : DataFrame
        Expected portfolio returns for each ticker and date
    """
    # TODO: Implement Function
#     exp_returns = lookahead_returns.copy()
#     for index, row in df_long.iterrows():
# #         print(abs(row-df_short.loc[index]))
#         n_row = ((row-df_short.loc[index]) * lookahead_returns.loc[index] ) / n_stocks
#         print(n_row)
#         exp_returns.loc[index] = n_row 
    
    
#     return exp_returns

    return ((df_long-df_short) * lookahead_returns) / n_stocks
project_tests.test_portfolio_returns(portfolio_returns)

Tests Passed


### View Data
Time to see how the portfolio did.

In [21]:
expected_portfolio_returns = portfolio_returns(df_long, df_short, lookahead_returns, 2*top_bottom_n)
print(expected_portfolio_returns.T.sum())
project_helper.plot_returns(expected_portfolio_returns.T.sum(), 'Portfolio Returns')

date
2013-07-31    0.00000000
2013-08-31    0.00000000
2013-09-30   -0.00236456
2013-10-31   -0.00222744
2013-11-30   -0.00096828
2013-12-31    0.00264429
2014-01-31   -0.00486972
2014-02-28   -0.00761406
2014-03-31   -0.00893045
2014-04-30   -0.00725501
2014-05-31   -0.00938457
2014-06-30   -0.00247228
2014-07-31    0.01026566
2014-08-31    0.01706372
2014-09-30    0.01259415
2014-10-31    0.03362185
2014-11-30    0.01201786
2014-12-31    0.01152313
2015-01-31   -0.01759709
2015-02-28    0.02254455
2015-03-31    0.00792112
2015-04-30    0.02539667
2015-05-31   -0.01203012
2015-06-30    0.03572643
2015-07-31   -0.00304893
2015-08-31    0.03952277
2015-09-30   -0.01182947
2015-10-31    0.01264570
2015-11-30   -0.00222172
2015-12-31   -0.01901051
2016-01-31    0.00987480
2016-02-29   -0.01067894
2016-03-31   -0.02589224
2016-04-30   -0.00242558
2016-05-31   -0.00056459
2016-06-30   -0.00255248
2016-07-31   -0.03299962
2016-08-31    0.00136873
2016-09-30    0.00989578
2016-10-31    0.0074

## Statistical Tests
### Annualized Rate of Return

In [22]:
expected_portfolio_returns_by_date = expected_portfolio_returns.T.sum().dropna()
portfolio_ret_mean = expected_portfolio_returns_by_date.mean()
portfolio_ret_ste = expected_portfolio_returns_by_date.sem()
portfolio_ret_annual_rate = (np.exp(portfolio_ret_mean * 12) - 1) * 100

print("""
Mean:                       {:.6f}
Standard Error:             {:.6f}
Annualized Rate of Return:  {:.2f}%
""".format(portfolio_ret_mean, portfolio_ret_ste, portfolio_ret_annual_rate))


Mean:                       0.003185
Standard Error:             0.002158
Annualized Rate of Return:  3.90%



The annualized rate of return allows you to compare the rate of return from this strategy to other quoted rates of return, which are usually quoted on an annual basis. 

### T-Test
Our null hypothesis ($H_0$) is that the actual mean return from the signal is zero. We'll perform a one-sample, one-sided t-test on the observed mean return, to see if we can reject $H_0$.

We'll need to first compute the t-statistic, and then find its corresponding p-value. The p-value will indicate the probability of observing a t-statistic equally or more extreme than the one we observed if the null hypothesis were true. A small p-value means that the chance of observing the t-statistic we observed under the null hypothesis is small, and thus casts doubt on the null hypothesis. It's good practice to set a desired level of significance or alpha ($\alpha$) _before_ computing the p-value, and then reject the null hypothesis if $p < \alpha$.

For this project, we'll use $\alpha = 0.05$, since it's a common value to use.

Implement the `analyze_alpha` function to perform a t-test on the sample of portfolio returns. We've imported the `scipy.stats` module for you to perform the t-test.

Note: [`scipy.stats.ttest_1samp`](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.stats.ttest_1samp.html) performs a two-sided test, so divide the p-value by 2 to get 1-sided p-value

In [26]:
from scipy import stats

def analyze_alpha(expected_portfolio_returns_by_date):
    """
    Perform a t-test with the null hypothesis being that the expected mean return is zero.
    
    Parameters
    ----------
    expected_portfolio_returns_by_date : Pandas Series
        Expected portfolio returns for each date
    
    Returns
    -------
    t_value
        T-statistic from t-test
    p_value
        Corresponding p-value
    """
    # TODO: Implement Function
    t_value,p_value = stats.ttest_1samp(expected_portfolio_returns_by_date,0)
    print(t_value,p_value)

    return t_value,p_value/2

project_tests.test_analyze_alpha(analyze_alpha)

-0.940764456618 0.416228196414
Tests Passed


### View Data
Let's see what values we get with our portfolio. After you run this, make sure to answer the question below.

In [27]:
t_value, p_value = analyze_alpha(expected_portfolio_returns_by_date)
print("""
Alpha analysis:
 t-value:        {:.3f}
 p-value:        {:.6f}
""".format(t_value, p_value))

1.47575494668 0.146678231359

Alpha analysis:
 t-value:        1.476
 p-value:        0.073339



### Question: What p-value did you observe? And what does that indicate about your signal?

*#TODO: Put Answer In this Cell*
0.073339 , it is above our alpha of 0.05 which means it passes our T-test ? 

## Submission
Now that you're done with the project, it's time to submit it. Click the submit button in the bottom right. One of our reviewers will give you feedback on your project with a pass or not passed grade. You can continue to the next section while you wait for feedback.